mimxrt: Provide WiFi and BLE support for MIMXRT boards.

robert-hh opened the pull request:
This support uses ESP32 modules with either NINAW10 firmware or esp-hosted firmware. For MIXRT101x devices, the NINAW10 firmware is expected. For MIMXRT102x, MIMXRT105x and MIMXRT106x the esp_hosted firmware has to be used. The esp_hosted firmware is faster and has a more robust BLE support, but the MIMXRT to support  it gets too larger for MIMXRT101x devices. This PR includes Pin definitions and enables WiFI and BLE for:

- ADAFRUIT_METRO_M7_AIRLIFT
- MIMXRT1010_EVK
- MIMXRT1015_EVK
- OLIMEX RT1010py
- MIMXRT1020_EVK
- MIMXRT1050_EVK
- Seeed Arch Mix
- Teensy 4.0
- Teensy 4.1

As ESP32 device, e.g. Adafruit Airlift modules, Nina W10 modules or any generic ESP32 breakout can be used. The NINAW10 or esp_hosted firmware has to be installed to that device. Therefore there are dependencies of this device to a few other PR's to micropython-lib, especially https://github.com/micropython/micropython-lib/pull/642 and https://github.com/micropython/micropython-lib/pull/654. The PR https://github.com/micropython/mynewt-nimble/pull/2 prevents a rare lock-up situation in the ble-nimble stack, which occurred during testing with the BLE test suite.

robert-hh committed:
mimxrt/BLE: Change nina_bt_hci.c to support RTS and CTS.

RTS must be low for operation.
Allow CS as alias for GPIO1.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/WLAN: Support external NINA10 modules for WLAN.

Using the network.WLAN module with Pin settings defined in
mpconfigboard.h.

To allow for port specific WiFi configurations, the functions in
drivers/ninaw10/nina_wifi_bsp.c are declared as MP_WEAK.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/WLAN: Adapt the mimxrt port files.

Change Makefile and mpconfigport.h.
Add a port specific nina_wifi_bsp.c file, which just increases the
poll rate.

The flag for enabling NinaW10 is MICROPY_PY_NETWORK_NINAW10.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/WLAN: Add Pin definitions for MIMXRT Boards.

- MIMXRT1010_EVK
- MIMXRT1015_EVK
- ADAFRUIT Metro M7

For the *_EVK boards the pins are those used by the Adafruit Airlift
shield.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/BLE: Change main.c, pendsv.h, mpconfigport.h, Makefile for BLE.

And reduce the stack size for BLE.

Nibmle uses 13k of static data, increasing the .bss section. Moving the
stack or bss to OCRM is possible, but reduces the Python heap size.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/NINAW10: Add or change board configuration files.

For:
MIMXRT1010_EVK
MIMXRT1015_EVK
ADAFRUIT Metro M7
OLIMEX RT1010

Adafruit Metro M7 board:

Add the SD variant of the Adafruit Metro M7 board.
It supports WIFI/BLE as well, but at pins matching the Airlift UNO
breakout.

Olimex 1010 board:

The pins are selected in a way that they are accessible at the UEXT1
connector of the Dev board.

There is no free Pin for GPIO0. So GPIO0 must be connected with a
resistor to a Pin which is not used for FW upload and is used as
output during normal operation, e.g. SPI SCL.
The NINA FW configures GPIO0 as output during normal operation.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt: Combined files and changes for using the esp_hosted driver.

At first with the board files for MIMXRT1020 only.

MIMXRT1010 and MIMXRT1015 do not build due to insufficient RAM size.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/esp_hosted: Add more esp_hosted boards files.

- Teensy 4.0
- Teensy 4.1
- MIMXRT1050_EVK
- SEEED Arch Mix

WiFi & BLE work for Teensy 4.x and SEEED Arch Mix with the given wiring.
The Teensy 4.x wiring matches the combination of a Adafruit Feather
Airlift module with Teensy - Feather adapter.
Limited activity for MIMXRT1050, since the SPI wires are not connected.
BLE on MIMXRT1050EVK worked, with CS and MOSI of the airlift shield
hardwired to GND.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
drivers/esp_hosted: Shorten the start-up time for BLE.

Instead of waiting a fixed long time for absorbing ESP32 boot messages,
just wait a shorter time for no more data to arrive

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/docs: Extend the documentation for WLAN and BLE.

Inluding instructions for connecting external modules and uploading
the WLAN/BLE firmware.

Signed-off-by: robert-hh <robert@hammelrath.com>

iabdalkader commented:
I think you may have some really outdated changes, I didn't look at the whole PR but note that `mimxrt` already has *working* WiFi and BLE support using `nimble` and `cyw43`, I added it myself some time ago, so some of your changes duplicate things (for example see `mp_bluetooth_deinit()` in `main.c`, and port files in Makefile etc..), also `nimble.mk` includes the paths you added. Like I said it's already working, you just need to add support for new NICs/drivers (Nina/esp_hosted).

robert-hh commented:
Thanks for the hint and maybe it's the case. The history of this PR starts earlier than the inclusion of the cyw43 support. Since you looked at it already, can you be specific about the places were things are duplicated?

iabdalkader commented on ports/mimxrt/Makefile:
```diff
@@ -217,6 +247,13 @@ SRC_C += \
 	ticks.c \
 	tusb_port.c \
 
+ifeq ($(MICROPY_PY_BLUETOOTH),1)
+SRC_C += \
+	drivers/ninaw10/nina_bt_hci.c \
+    mpbthciport.c \
+    mpnimbleport.c
+endif
+
```
This is already included somewhere else in the Makefile, I'd add to it:

```C
ifeq ($(MICROPY_PY_BLUETOOTH),1)
SRC_C += mpbthciport.c
ifeq ($(MICROPY_PY_NETWORK_NINAW10),1)
SRC_C += drivers/ninaw10/nina_bt_hci.c
endif
endif # MICROPY_PY_BLUETOOTH
```

robert-hh replied:
That's settled now, I hope.

iabdalkader commented on ports/mimxrt/Makefile:
```diff
@@ -106,8 +122,9 @@ SRC_TINYUSB_C += \
-# All settings for Ethernet support are controller by the value of MICROPY_PY_LWIP
+# All settings for Ethernet support are controller by the value of MICROPY_PY_LWIP and the supported MCUs
 ifeq ($(MICROPY_PY_LWIP),1)
+ifeq ($(MCU_SERIES),$(filter $(MCU_SERIES), MIMXRT1021 MIMXRT1052 MIMXRT1062 MIMXRT1064 MIMXRT1176))
```
I don't know about this, prefer to control it via `MICROPY_PY_LWIP`, sometimes you need to disable it for a board.

robert-hh replied:
That's an old setting related to Ethernet. It might result to the same, since all these boards have LWIP enabled, and the MIMXRT1011 and MIMXRT1015 do not use LWIP.

iabdalkader commented on ports/mimxrt/Makefile:
```diff
@@ -233,6 +270,14 @@ SHARED_SRC_C += \
 	shared/runtime/sys_stdio_mphal.c \
 	shared/timeutils/timeutils.c \
 
+ifeq ($(MICROPY_PY_BLUETOOTH),1)
+SRC_C += mpbthciport.c
+endif
+
+ifeq ($(MICROPY_BLUETOOTH_NIMBLE),1)
+SRC_C += mpnimbleport.c
+endif
+
```
Those already exist in the file.
```suggestion
```

iabdalkader commented on ports/mimxrt/main.c:
```diff
@@ -55,6 +55,10 @@
 
 #include "systick.h"
 #include "extmod/modnetwork.h"
+#if MICROPY_PY_BLUETOOTH
+#include "extmod/modbluetooth.h"
+#include "mpbthciport.h"
+#endif
```
```suggestion
```

robert-hh replied:
Removed.

iabdalkader commented on ports/mimxrt/main.c:
```diff
@@ -66,6 +70,10 @@ int main(void) {
     tusb_init();
     pendsv_init();
 
+    #if MICROPY_PY_BLUETOOTH
+    mp_bluetooth_hci_init();
+    #endif
+
```
```suggestion
```

robert-hh replied:
done

iabdalkader commented on ports/mimxrt/main.c:
```diff
@@ -106,6 +113,15 @@ int main(void) {
         #if MICROPY_PY_NETWORK
         mod_network_init();
         #endif
+        #if MICROPY_PY_BLUETOOTH
+        mp_bluetooth_hci_init();
+        #endif
```
```suggestion
```

robert-hh replied:
removed,

dpgeorge replied:
I think this still needs to be moved back up outside the soft reset loop, where it used to be.

iabdalkader commented on ports/mimxrt/main.c:
```diff
@@ -106,6 +113,15 @@ int main(void) {
+        mp_bluetooth_hci_init();
+        #endif
+
+        #if MICROPY_PY_LWIP
+        // mod_network_lwip_init();
+        systick_enable_dispatch(SYSTICK_DISPATCH_LWIP, mod_network_lwip_poll_wrapper);
+        #endif
```
Why disable/enable the dispatch on soft-reboot ?

robert-hh replied:
The idea was to ensure, that during shutdown no events are handled.

iabdalkader commented on ports/mimxrt/main.c:
```diff
@@ -145,12 +161,22 @@ int main(void) {
         #if MICROPY_PY_NETWORK
         mod_network_deinit();
         #endif
+        #if MICROPY_PY_BLUETOOTH
+        mp_bluetooth_deinit();
+        #endif
```
```suggestion
```

robert-hh replied:
removed the duplicate

iabdalkader commented on ports/mimxrt/mpconfigport.h:
```diff
@@ -122,18 +124,29 @@ uint32_t trng_random_u32(void);
-#define MICROPY_PY_WEBSOCKET                (MICROPY_PY_LWIP)
-#define MICROPY_PY_WEBREPL                  (MICROPY_PY_LWIP)
+
+#ifndef MICROPY_PY_SSL
+#define MICROPY_PY_SSL                      (0)
+#endif
+#define MICROPY_PY_WEBSOCKET                (MICROPY_PY_LWIP || MICROPY_PY_NETWORK_NINAW10)
+#define MICROPY_PY_WEBREPL                  (MICROPY_PY_LWIP || MICROPY_PY_NETWORK_NINAW10)
```
Those are just enabled unconditionally in other ports, I think it's okay now to enable them here.

robert-hh replied:
It's not enabled yet for all boards. I could not test the MIMXRT1060 and 1064 boards. And did not test the MIMXRT1176 board yet.

iabdalkader commented on ports/mimxrt/mpconfigport.h:
```diff
@@ -177,21 +191,44 @@ extern const struct _mp_obj_type_t network_lan_type;
 #define MICROPY_HW_NIC_CYW43
 #endif
 
+#if MICROPY_PY_NETWORK_ESP_HOSTED
+extern const struct _mp_obj_type_t mod_network_esp_hosted_type;
+#define MICROPY_HW_NIC_ESP_HOSTED   { MP_ROM_QSTR(MP_QSTR_WLAN), MP_ROM_PTR(&mod_network_esp_hosted_type) },
+#else
+#define MICROPY_HW_NIC_ESP_HOSTED
+#endif
```
I would use `MICROPY_BOARD_NETWORK_INTERFACES` to add the new interfaces.

robert-hh replied:
But that would required to add it to every supported board, instead making it depending on MICROPY_PY_NETWORK_ESP_HOSTED

iabdalkader commented:
> Since you looked at it already, can you be specific about the places were things are duplicated?

I left a quick review, anything removed I think is duplicated.

> Edit2: I know the there is support for cyw43. Only the existing boards cannot use it. The added support for NINA or esp_hosted based devices allow users to add WiFi/BLE to other boards as well.

I understand, but what I meant is: there's already working support for `nimble` and `LWIP`, in that port, which means you only need to add support for Nina/Esp hosted.

iabdalkader commented on ports/mimxrt/Makefile:
```diff
@@ -81,6 +89,14 @@ INC += -I$(TOP)/lib/tinyusb/src
 INC += -I.
 INC += -Ihal
 
+ifeq ($(MICROPY_PY_BLUETOOTH),1)
+INC += \
+    -I$(TOP)/extmod/nimble \
+    -I$(TOP)/lib/mynewt-nimble/nimble/host/include \
+    -I$(TOP)/lib/mynewt-nimble/nimble/include \
+    -I$(TOP)/lib/mynewt-nimble/porting/nimble/include
+endif
+
```
This is already included by `nimble.mk`
```suggestion
```

robert-hh replied:
I removed it.

robert-hh committed:
mimxrt: Clean up Makefile, MIMXRT1015.ld and board files.

- Makefile, main.c: Remove duplicated code and settings, which were added
  through combining several branches.
- Install the proto-c compiler in ci.sh.
- Move _dcd_data to the start of the .bss segment. This data item has
  to be 2k aligned. Moving it to the start of the segment avoids
  wasting memory.
- Remove the board's manifest.py, now that many boards support
  networking.
- Remove obsolete settings from mpconfigboard.mk files.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
Thanks for the review. I was a little bit overwhelmed by the fact that git managed to merge 3 branches with almost no conflict. But I did not check for duplicated. Still CI reports a build error: `[proto_c cannot be found](make: protoc-c: Command not found)`. Sometimes it works.

iabdalkader commented:
> Thanks for the review. I was a little bit overwhelmed by the fact that git managed to merge 3 branches

Yes it looked like a bad git merge. Anyway, I can take another look tomorrow.

> Still CI reports a build error: `[proto_c cannot be found](make: protoc-c: Command not found)`. Sometimes it works.

You must `make submodules` with the BOARD set, see `ci_renesas_ra_board_build` in `tools/ci.sh`

robert-hh commented:
About Proto-c: It works locally. And when I do the make BOARD=xxx submodules and try to push it, there is no change to push.

iabdalkader commented:
```
function ci_mimxrt_build {
    make ${MAKEOPTS} -C mpy-cross
    make ${MAKEOPTS} -C ports/mimxrt BOARD=MIMXRT1020_EVK submodules
    make ${MAKEOPTS} -C ports/mimxrt BOARD=MIMXRT1020_EVK
    make ${MAKEOPTS} -C ports/mimxrt BOARD=TEENSY40 submodules
    make ${MAKEOPTS} -C ports/mimxrt BOARD=TEENSY40
}
```

I'm not sure why boards are built twice, but I think you need to add submodules to every build.

iabdalkader commented:
Oh okay first line checks out the submodules, in that case I'm not sure why it's failing.

robert-hh commented:
Proto-c was not installed. See a few lines up. The error about proto-c is now gone, only I have to clean up the manifest.py files.

robert-hh commented:
Closing in favor of supporting instead PPP. Then the user can go for a communication suitable method.

andrewleech commented:
Hi @robert-hh other than the chain of dependencies of this PR is there anything particularly wrong with it? 
Are there conflicts with the PPP work?

For context in planning on using this PR at work to add wireless networking to mimxrt platforms, PPP doesn't really help here at all. 

I don't have suitable hardware set up though so haven't scheduled testing this yet.

robert-hh commented:
> Hi @robert-hh other than the chain of dependencies of this PR is there anything particularly wrong with it?

No. It works fine. And there should be not conflict with PPP, other than lacking PPP would not work on MIMXRT101x boards. If you think it's useful, I can reopen it.

robert-hh commented:
@dpgeorge Thank you for your response. I had a PR for similar support at the SAMD port. I closed it, expecting that with the VFSROM support the flash space would not be sufficient any more. The WiFI only version leaves ~90k free, the WiFi+BLE variant about 5k.

dpgeorge commented:
Let's get this one through first, then we can consider the samd version.

dpgeorge commented on drivers/ninaw10/nina_wifi_bsp.c:
```diff
@@ -46,67 +46,82 @@
+#define MICROPY_HW_NINA_CS MICROPY_HW_NINA_GPIO1
+#endif
+
+MP_WEAK int nina_bsp_init(void) {
```
I don't think any of these `MP_WEAK` are needed.  It seems that the only function that's overridden by mimxrt is `nina_bsp_spi_slave_select`, and the only difference I could see there is that `mp_hal_delay_ms(1)` is replaced with `mp_hal_delay_us(100)`.  Is that change necessary?  If so, maybe it should be changed to `mp_hal_delay_us(100)` for everyone?

robert-hh replied:
Testing performance with the 1ms delay shows a drop of 15-20% vs the 100µs delay. That might not be worth the change. WiFi is slow using the NINA firmware, but at least it works.

dpgeorge commented on drivers/ninaw10/nina_wifi_bsp.c:
```diff
@@ -46,67 +46,82 @@
+// Use the name CS instead of GPIO1
+#ifndef MICROPY_HW_NINA_CS
+#define MICROPY_HW_NINA_CS MICROPY_HW_NINA_GPIO1
+#endif
```
I feel like it's not worth having this alias.  Just leave it as GPIO1.

(I'm trying to get minimal changes to this driver, to make this PR easier to understand, and less chance to break existing users of ninaw10.)

dpgeorge replied:
OK, makes sense.

Maybe it's worth putting the above table in https://github.com/micropython/micropython-lib/pull/642 ?

dpgeorge commented on ports/mimxrt/boards/ADAFRUIT_METRO_M7/mpconfigboard.h:
```diff
@@ -58,15 +58,39 @@
 #define MICROPY_HW_BLE_UART_BAUDRATE (115200)
+#define MICROPY_HW_BLE_UART_FLOW_CONTROL (0)
+
+#if MICROPY_HW_WIFI_AIRLIFT
```
I don't understand the Airlift vs non-Airlift build.  It just looks like a difference in pin assignments for the Nina?  Why does there need to be two build variants, why not just always use the on-board ESP32?

robert-hh replied:
The intention was to allow using the Airlift shield with the AIrlift-SD board. Then the signals are placed at other pins.

robert-hh committed:
drivers/ninaw10: Use the nina_wifi_bsp.c file of the NINAW10 driver.

And remove the MP_WEAK modifiers from that file, as suggested in the
review.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented on tools/ci.sh:
```diff
@@ -295,6 +295,7 @@ function ci_webassembly_run_tests {
 
 function ci_mimxrt_setup {
     ci_gcc_arm_setup
+    sudo apt-get install protobuf-c-compiler
```
Please also add the same line to `ci_code_size_setup`, which builds TEENSY40.

dpgeorge commented on drivers/ninaw10/nina_wifi_bsp.c:
```diff
@@ -46,45 +46,56 @@
     mp_hal_pin_write(MICROPY_HW_NINA_RESET, 1);
     mp_hal_delay_ms(750);
 
-    mp_hal_pin_write(MICROPY_HW_NINA_GPIO0, 0);
```
I think this line needs to stay, and put within the #ifdef below.

dpgeorge commented on drivers/ninaw10/nina_wifi_bsp.c:
```diff
@@ -103,7 +114,11 @@ int nina_bsp_atomic_exit(void) {
 }
 
 int nina_bsp_read_irq(void) {
```
I don't think this function is ever used, so maybe just wrap the whole thing in `#ifdef MICROPY_HW_NINA_GPIO0`.

dpgeorge commented on ports/mimxrt/mpconfigport.h:
```diff
@@ -57,7 +57,9 @@ uint32_t trng_random_u32(void);
 #define MICROPY_STACK_CHECK_MARGIN          (1024)
 #define MICROPY_ENABLE_EMERGENCY_EXCEPTION_BUF  (1)
 #define MICROPY_LONGINT_IMPL                (MICROPY_LONGINT_IMPL_MPZ)
+#define MICROPY_SCHEDULER                   (1)
```
This is not a config option, so can be removed.

robert-hh replied:
Gone.

dpgeorge commented on ports/mimxrt/mpconfigport.h:
```diff
@@ -57,7 +57,9 @@ uint32_t trng_random_u32(void);
 #define MICROPY_LONGINT_IMPL                (MICROPY_LONGINT_IMPL_MPZ)
+#define MICROPY_SCHEDULER                   (1)
 #define MICROPY_SCHEDULER_DEPTH             (8)
+#define MICROPY_ENABLE_SCHEDULER            (1)
```
This should already be enabled at the selected MICROPY_CONFIG_ROM_LEVEL.  So no need to explicitly enable it (to keep config at a minimum).

dpgeorge commented on ports/mimxrt/mpconfigport.h:
```diff
@@ -183,13 +196,44 @@ extern const struct _mp_obj_type_t network_lan_type;
+
+#if MICROPY_PY_NETWORK_CYW43
+extern const struct _mp_obj_type_t mp_network_cyw43_type;
+#define MICROPY_HW_NIC_CYW43                { MP_ROM_QSTR(MP_QSTR_WLAN), MP_ROM_PTR(&mp_network_cyw43_type) },
```
As above, not needed.

dpgeorge reviewed (commented):
Hi @robert-hh I'm finally coming back to this PR :sweat_smile:

If you still think it's worth merging, and you have the time to update it, I've made some comments.

Overall it looks pretty good.  Because all the features already exist, this PR is mostly about wiring things up, adding glue code, and setting configuration.

robert-hh commented:
> If you still think it's worth merging, and you have the time to update it,

Thanks for picking this up. I have time to update it. But I'm not sure about the PR's value. I thought it would when I started the work. But it adds code, that will need maintenance sometime. The only board that supports WiFi is the Adafruit Metro M7. All other boards I know need an add-on. These are available a.t.m., even for Teensy, but it is a little bit of work, and I do not know how many people would do that. So I'm not sure looking at the balance between benefit and burden.

robert-hh committed:
mimxrt_wifi_ble: Implement the review comments.

Mostly removing duplicate or obsolete settings.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/boards: Enable/disable WiFi/BLE with a single switch.

Allowing to disable WiFi and BLE by the value of
MICROPY_PY_NETWORK_ESP_HOSTED in a board's  mpconfigboard.mk.

Remove duplicate lines from Makefile.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge committed:
tools/ci.sh: Include mynewt-nimble in submodules for code size job.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tools/ci.sh: Include protobuf-c in submodules for code size job.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented on ports/mimxrt/Makefile:
```diff
@@ -322,14 +357,6 @@ else
     SRC_HAL_C += hal/$(FLEXSPI_FLASH_TYPE)_config.c
 endif
 
-ifeq ($(MICROPY_PY_BLUETOOTH),1)
```
Unless there was a reason to move these two bits up in the file, I suggest keeping them where they were (ie not removing them from here and not adding them above).

robert-hh committed:
mimxrt/wlan_pinout and mimxrt/Makefile: Implement review comments.

docs/mimxrt/wlan_pinout.rst: Refer to NINA_FW version 2.0
ports/mimxrt/Makefile: Move config lines back to where they were.

side change: Fix two comments at Teensy mpconfigboard.h and change
the initial baudrate for MIMXRT1050_EVK.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
I ran a speed comparison between the esp_hosted versions and the NINA FW. The comparison may be somewhat unfair because it ran at a slower board with less RAM. SPI baud rate is 8MHz, which is the hard upper limit for transfer speed. Besides ping, the esp_hosted fw versions perform close.

```
Test                ---- esp_hosted ---  NINA FW
                    ---- Teensy 4.0 ---  Metro M7
                    V0.0.5   V1.0.0.0.0  V2.0
-------------------------------------------------
iperf3 TCP (MBit/s)  3.36      3.60      1.19
iperf3 UDP (MBit/s)  6.26      6.28      0.57
xmit_64k (MBit/s)    3.41      3.59      1.59
ping (ms)            86        13        66
```

iabdalkader commented:
There's been so many notifications from this PR, so I haven't been following and missed mentions.
@robert-hh it sounds like you got it working? So the driver working with `v1.0.0.0.0`? Great job! A few months ago, I saw that they made some changes to the protocol for FG, that I thought might be breaking changes. It's very good to hear that the driver works with the latest firmware.

Could you share the build instructions? We're also interested in running it on the Arduino Nano board that runs Nina firmware?

It's probably worth mentioning that I added an esp-hosted driver to Zephyr a while ago, here: https://github.com/zephyrproject-rtos/zephyr/pull/87685. It turns out there a lightweight replacement for protobuf, [nanopb](https://github.com/nanopb/nanopb), that can allocate most things statically, and has a more permissive license, I think, compared to protobuf. So just keep this in mind, there's room to improve the licensing and footprint of the MicroPython driver. (CC @dpgeorge).

robert-hh commented:
@iabdalkader Thanks. Attached are the sdkconfig files which I used for V1.x. The SPI Pin config is for Airlift boards. But that is easy changeable. 
[sdkconfig_V1_000.zip](https://github.com/user-attachments/files/24106553/sdkconfig_V1_000.zip)

iabdalkader commented:
> It may still be interesting to tell, why the v1.0.0.0 version of esp_hosted behaves strange

@robert-hh I got the MicroPython driver to work with both the v0.0.5 and v1.0.0.0.0. I did basic testing but both BLE and WiFi are working fine, I see performance improvements as well (probably due to bigger LWIP buffer I used). I will be sending patches to add support for this later.

>  Edit: Already building using that path seems to be a rough ride.

See this new repo I created here: https://github.com/iabdalkader/esp-hosted-firmware this has a workflow that fetches esp-idf, esp-hosted etc.. applies patches, copies sdkconfig, builds the firmware and creates a release on tag pushes. Please feel free to send PRs to add more targets if you like. To add a target you'll need to add it its config to the workflow (pretty straightforward), add your sdkconfig, and patches (if any). Note you can test the generated binary from workflow artifacts if you do send PRs.
